### DISCUSSION TITLE
use string as flags

### DISCUSSION THREAD

**Discussion by The-Lost-Light** [nushell/nushell#13474] [Q&A] [answered]
I've been searching for a long time but still haven't found a solution.
Can i use a string in a variable and pass it as a flag(with value) to a command
```nu
def main [--keyboard (-k)] {
	let keyboard_flag = if $keyboard { "-d *::kbd_backlight" } else { "" }
	brightnessctl get $keyboard_flag
}
```
I want it like this `brightnessctl get -d *::kbd_backlight`
And it actually is `brightnessctl get "-d *::kbd_backlight"`

**Comment by ghost** [comment]
yes flags can contain values: just specify the datatype. example:

```nu
def [
  --verbose(-v)
  --name: string  # btw you can put descriptions (aka help) here as comments
  --port: int
] { }
```

**Reply by The-Lost-Light** [reply]
You may misunderstand my idea.
Can a string to be flag and the value of flag. (not the parameters of script)
like this: 
```nu
let keyboard_flag = "-d *::kbd_backlight"
brightnessctl get $keyboard_flag
```
or i have to separate the flag by my self like
```nu
let keyboard_flag = "*::kbd_backlight"
brightnessctl get -d $keyboard_flag
```
After all, it is a bit complicated to deal with whether to add a flag. (consider value is empty then i dont need -d)
It would be much more convenient if the flag and the value of the flag could be passed directly at once.

**Reply by ghost** [reply]
so you want dynamic flags, which get defined within the function itself?  
as far as im aware this is not possible (and would mess with completion, help, lsp, etc).  
you could use `[...wargs: string]`, add a own parser and let the user wrap flags in quotes, but thats probably to much workaround.  
a better solution would probably be to either just add all flags and ignore them per default, setting the base-flag type to dictionary  (or list or whatever), or use subcommands (`def "main keyboard" [] {}`, `def "main mouse" [] {}`, etc).

> After all, it is a bit complicated to deal with whether to add a flag. (consider value is empty then i dont need -d)

if you add all flags you could put a `assert` in there to verify it (weather at the position it gets used or beginning is up to you) or add default values

**Reply by The-Lost-Light** [reply]
Ok I took some time to test
```nu
let flags = {--device: "*::kbd_backlight", --foo: "foo"}
brightnessctl get (if ($flags.--device | is-not-empty) {"-d"}) $flags.--device
```
I'm still a little confused
1. Is there any method that i can get the index of record like `{a:1,b:2} | find-key-by-value 1` then i get the key "a" ("where" can't do this) or `{a:"abc",b:"", c:"def"} | key-if-value-not-empty` then get ["a", "c"]
2. If i have a lots of flags  have to list them all manually, there is no way to add them automatically.
3. I find the "run-external" command, but the function is similar to the common execution method, why it can to designed be pass in a list as parameter like `run-external ["'command' '-f1' 'v1' '-f2'"]` or even pass in a string directly

Finally, in my case, is run `sh -c` directly a better choice?
```nu
let flag = "-d *::kbd_backlight"
sh -c $"brightnessctl get ($flag)"
```

**Reply by ghost** [reply, marked as answer]
oh, sorry i still thought you wanted to change the flags of the nu-function while executing it and not that you wanted to build a external command.

for external commands you can use the spread operator (`...`) on a list:
```nu
systemctl ...["--user", "status"]
```

and while it used to be possible in the past to just `[(if $foo {"--foo"}]` you now need to always define a else `[(if $foo {"--foo"} else {""})]`. but since this would add empty (`""`) arguments we can use the spread operator again: `[...(if $foo {["--foo"] else []})]`.  
a full example could look like this:

```nu
^my_external_command ...[
  "status"
  ...(if $user {["--user"]} else {[]})
  ...(if $foo != null {["--foo", $foo] else {[]}}
]
```

**Reply by The-Lost-Light** [reply]
Thanks this solved my problem perfectly
and I found a short way~
```nu
let flags =  [(if $keyboard { [--device *::kbd_backlight] })] | flatten | compact
brightnessctl get ...$flags
```
